sw/device/tests/sram_ctrl_execution_test_ret.c

```diff
@@ -0,0 +1,175 @@
+
+static const uint32_t kRamStartAddr =
+    TOP_EARLGREY_SRAM_CTRL_RET_AON_RAM_BASE_ADDR;
+static const uint32_t kRamEndAddr =
+    TOP_EARLGREY_SRAM_CTRL_RET_AON_RAM_BASE_ADDR +
+    TOP_EARLGREY_SRAM_CTRL_RET_AON_RAM_SIZE_BYTES - 1;
+
+/*
+ * Buffer to hold instructions to be executed.
+ *
+ * This can be thought of as the implementation for the `ram_exec_function_t`.
+ */
+static const uint32_t execution_test_instructions[2] = {
+    /*
+     *   Abstract representation: add rd, rs1, rs2
+     *   +-----------+-------+-------+-------+------+-----------+
+     *   |  0000000  |  rs2  |  rs1  |  000  |  rd  |  0110011  |
+     *   +-----------+-------+-------+-------+------+-----------+
+     *   31        25 24   20 19   15 14   12 11   7 6          0
+     *
+     *   Custom representation: add a0, a1, a0
+     *   +-----------+---------+---------+-------+--------+-----------+
+     *   |  0000000  |  01010  |  01011  |  000  | 01010  |  0110011  |
+     *   +-----------+---------+---------+-------+--------+-----------+
+     *   31        25 24     20 19     15 14   12 11     7 6          0
+     */
+    0x00A58533,
+    /*
+     *   Abstract representation: ret == jalr rd, offset(rs1)
+     *   +----------------+-------+-------+------+-----------+
+     *   |  000000000000  |  rs1  |  000  |  rd  |  1100111  |
+     *   +----------------+-------+-------+------+-----------+
+     *   31             20 19   15 14   12 11   7 6          0
+     *
+     *   Custom representation: jalr zero, 0(ra)
+     *   +----------------+---------+-------+---------+-----------+
+     *   |  000000000000  |  00001  |  000  |  00000  |  1100111  |
+     *   +----------------+---------+-------+---------+-----------+
+     *   31             20 19     15 14   12 11      7 6          0
+     */
+    0x00008067,
+};
+
+/*
+ * Handles an exception.
+ *
+ * This exception handler keeps default exception implementations for faults
+ * that are not relevant in terms of this test. The only fault handler that
+ * is overridden is Instruction Access Fault.
+ *
+ * The controlled Instruction Access Fault occurs originates in the retention
+ * SRAM, which means that normally the return address would be calculated
+ * relevant to the offending instruction. However, due to execution from
+ * retention SRAM being permanently disabled, this approach would not work.
+ * The return address for the one stack frame before is used instead. In
+ * other words sram_execution_test -> retention_sram -> exception_handler
+ * -> sram_execution_test.
+ *
+ * Before the jump into the exception handler, the register set is saved on
+ * stack. This means that the return address can be loaded from the stack, or
+ * simply read from the return address register (before it is overridden by
+ * the code in this exception handler). This means that the return address
+ * register value must be read before any function call.
+ */
+void handler_exception(void) {
```

Comment:
Just an FYI: overriding this function is ok for now, since you need to save the return address. IIUC, the [`INTERRUPT_HANDLER_ABI`](https://github.com/lowRISC/opentitan/blob/1ac30bea3b2502e46137e30794349e5452618bf2/sw/device/lib/handler.h#L35) attribute is applied to this function prototype in `sw/device/lib/handler.h` so the compiler should save `ra` to the stack before the  `fault_handler_testutils_set_return_addr()` function is invoked. So theoretically you could pluck it off the stack in `fault_handler_testutils_set_return_addr()`. However, IIRC, the compiler will only save those registers that are clobbered in this function to the stack, so it's hard to guess (and maintain) the stack offset where the `ra` will be.

When we migrate this test to use the OTTF however, this will not be the case, since the [OTTF saves **_all_** registers to the stack on an exception entry](https://github.com/lowRISC/opentitan/blob/0310540ed5e4e25c3b695db326027726ff5b434f/sw/device/lib/testing/test_framework/freertos_port.S#L24). In that case, I think you should know the exact stack offset, and should be able to do as I describe above.

Additionally, when we migrate this test to the OTTF, I think you will just want to override the [`ottf_instr_access_fault()`](https://github.com/lowRISC/opentitan/blob/1ac30bea3b2502e46137e30794349e5452618bf2/sw/device/lib/testing/test_framework/ottf_isrs.c#L54) function (since its tagged as "weak"). Then, we can eliminate the `fault_handler_testutils_set_return_addr()` test utils function (since half of that test util, that deals with updating the `mepc`, is already implemented by the OTTF ISR asm).

To be clear, this is ok for now, until we migrate to the OTTF. Just wanted to give you a heads up :)